src/bentoml/_internal/monitoring/api.py

```diff
@@ -0,0 +1,302 @@
+from __future__ import annotations
+
+import typing as t
+import logging
+import datetime
+import contextlib
+import collections
+import logging.config
+from typing import TYPE_CHECKING
+
+import yaml
+from simple_di import inject
+from simple_di import Provide
+
+from ..types import LazyType
+from ..context import component_context
+from ..configuration.containers import BentoMLContainer
+
+if TYPE_CHECKING:
+    from ..types import JSONSerializable
+
+DT = t.TypeVar("DT")
+MT = t.TypeVar("MT", bound="Monitor[t.Any]")
+
+logger = logging.getLogger(__name__)
+
+BENTOML_MONITOR_ROLES = {"feature", "prediction", "target"}
+BENTOML_MONITOR_TYPES = {"numerical", "categorical", "numerical_sequence"}
+
+MONITORS: dict[str, Monitor[t.Any]] = {}  # cache of monitors
+
+
+class Monitor(t.Generic[DT]):
+    def __init__(self, name: str):
+        pass
+
+    def start_record(self) -> None:
+        pass
+
+    def stop_record(self) -> None:
+        pass
+
+    def export_schema(self) -> JSONSerializable:
+        pass
+
+    def export_data(self) -> JSONSerializable:
+        pass
+
+    def log(self, data: DT, name: str, role: str, data_type: str) -> None:
+        pass
+
+    def log_batch(
+        self, data_batch: t.Iterable[DT], name: str, role: str, data_type: str
+    ) -> None:
+        pass
+
+    def log_table(
+        self,
+        data: t.Iterable[t.Iterable[DT]],
+        schema: dict[str, str],
+    ) -> None:
+        pass
```

Comment:
should this section be a abstractmethod? I feel like all Monitor class should implement all of these.

Reply:
done